ext/semian_cb_data/semian_cb_data.c

```diff
@@ -0,0 +1,900 @@
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (0 == ptr->shm_address)
+    return Qnil;
+
+  if (!semian_cb_data_lock(self))
+    return Qnil;
+
+  VALUE retval;
+  if (ptr->shm_address->arr_length > 0)
+    retval = DBL2NUM(ptr->shm_address->errors[ptr->shm_address->arr_length-1]);
+  else
+    retval = Qnil;
+
+  semian_cb_data_unlock(self);
+  return retval;
+}
+
+void
+Init_semian_cb_data (void) {
+
+  VALUE cSemianModule = rb_const_get(rb_cObject, rb_intern("Semian"));
+
+  VALUE cCircuitBreakerSharedData = rb_const_get(cSemianModule, rb_intern("CircuitBreakerSharedData"));
+
+  rb_define_alloc_func(cCircuitBreakerSharedData, semian_cb_data_alloc);
+  rb_define_method(cCircuitBreakerSharedData, "_initialize", semian_cb_data_init, 3);
+  rb_define_method(cCircuitBreakerSharedData, "_cleanup", semian_cb_data_clean, 0);
+  //rb_define_method(cCircuitBreakerSharedData, "acquire_semaphore", semian_cb_data_acquire_semaphore, 1);
+  //rb_define_method(cCircuitBreakerSharedData, "delete_semaphore", semian_cb_data_delete_semaphore, 0);
+  //rb_define_method(cCircuitBreakerSharedData, "lock", semian_cb_data_lock, 0);
+  //rb_define_method(cCircuitBreakerSharedData, "unlock", semian_cb_data_unlock, 0);
+  //rb_define_method(cCircuitBreakerSharedData, "acquire_memory", semian_cb_data_acquire_memory, 1);
+  //rb_define_method(cCircuitBreakerSharedData, "delete_memory", semian_cb_data_delete_memory, 0);
+
+  rb_define_method(cCircuitBreakerSharedData, "semid", semian_cb_data_semid, 0);
+  rb_define_method(cCircuitBreakerSharedData, "shmid", semian_cb_data_shmid, 0);
+  rb_define_method(cCircuitBreakerSharedData, "successes", semian_cb_data_get_successes, 0);
+  rb_define_method(cCircuitBreakerSharedData, "successes=", semian_cb_data_set_successes, 1);
+
+  rb_define_method(cCircuitBreakerSharedData, "[]", semian_cb_data_array_at_index, 1);
+  rb_define_method(cCircuitBreakerSharedData, "[]=", semian_cb_data_array_set_index, 2);
+  rb_define_method(cCircuitBreakerSharedData, "length", semian_cb_data_array_length, 0);
+  rb_define_method(cCircuitBreakerSharedData, "size", semian_cb_data_array_length, 0);
+  rb_define_method(cCircuitBreakerSharedData, "count", semian_cb_data_array_length, 0);
+  rb_define_method(cCircuitBreakerSharedData, "<<", semian_cb_data_set_push_back, 1);
+  rb_define_method(cCircuitBreakerSharedData, "push", semian_cb_data_set_push_back, 1);
+  rb_define_method(cCircuitBreakerSharedData, "pop", semian_cb_data_set_pop_back, 0);
+  rb_define_method(cCircuitBreakerSharedData, "shift", semian_cb_data_set_pop_front, 0);
+  rb_define_method(cCircuitBreakerSharedData, "unshift", semian_cb_data_set_push_front, 1);
+  rb_define_method(cCircuitBreakerSharedData, "clear", semian_cb_data_array_clear, 0);
+  rb_define_method(cCircuitBreakerSharedData, "first", semian_cb_data_array_first, 0);
+  rb_define_method(cCircuitBreakerSharedData, "last", semian_cb_data_array_last, 0);
```

Comment:
Array [] is not needed, but `#<<`, `#shift`, `#first`, `#size` are in `Semian::CircuitBreaker#push_time`, `#last` is in `Semian::CircuitBreaker#error_timeout_expired`, `#clear` is in `Semian::CircuitBreaker#reset` and `Semian::CircuitBreaker#close`, and the rest were provided for completeness. I can remove some of the unused ones.